Move the CUDA implementation of log1p to ATen.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28237 Migrate `sin` and `sin_` from the TH to Aten (CUDA)
* #27372 Move the CUDA implementation of sqrt to ATen.
* **#26923 Move the CUDA implementation of log1p to ATen.**


Benchmark (RHEL 7, gcc 8.3.1, P1000):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.log1p(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.log1p(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")/100000', number=t))
```

Before:

```
torch.log1p(a) a.numel() == 10000 for 20000 times torch.half
0.46644441100033873
torch.log1p(a) a.numel() == 10000 for 20000 times torch.float
0.47403449599914893
torch.log1p(a) a.numel() == 10000 for 20000 times torch.double
0.5681769420007186
torch.log1p(a) a.numel() == 100000 for 20000 times torch.half
0.5420387039994239
torch.log1p(a) a.numel() == 100000 for 20000 times torch.float
0.6156843030003074
torch.log1p(a) a.numel() == 100000 for 20000 times torch.double
2.4580643359986425
```

After:

```
torch.log1p(a) a.numel() == 10000 for 20000 times torch.half
0.46042340799976955
torch.log1p(a) a.numel() == 10000 for 20000 times torch.float
0.4593915469995409
torch.log1p(a) a.numel() == 10000 for 20000 times torch.double
0.559550654999839
torch.log1p(a) a.numel() == 100000 for 20000 times torch.half
0.513594127000033
torch.log1p(a) a.numel() == 100000 for 20000 times torch.float
0.6011945249993005
torch.log1p(a) a.numel() == 100000 for 20000 times torch.double
2.444039010999404
```

Fix #24588

Differential Revision: [D17984184](https://our.internmc.facebook.com/intern/diff/D17984184)

xuhdev committed:
Move the CUDA implementation of log1p to ATen.

Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
LGTM, Please rebase and I will start landing.

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

[ghstack-poisoned]

xuhdev commented:
Rebased

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

Differential Revision: [D17984184](https://our.internmc.facebook.com/intern/diff/D17984184)

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

Differential Revision: [D17984184](https://our.internmc.facebook.com/intern/diff/D17984184)

[ghstack-poisoned]

xuhdev committed:
Update on "Move the CUDA implementation of log1p to ATen."


Fix #24588

Differential Revision: [D17984184](https://our.internmc.facebook.com/intern/diff/D17984184)

[ghstack-poisoned]
